### ISSUE TITLE
predicted bounding boxes deviate significantly from the ground-truth annotations

### ISSUE THREAD

**Issue by shenxhui** [Sompote/SAM3_LoRA#19]
Great work！Thank you for open-sourcing this project!

I fine-tuned the lora model on my own dataset and experimented with multiple hyperparameter settings. However, the predicted bounding boxes deviate significantly from the ground-truth annotations.

I noticed that in the training code, the coordinate format is converted from (x, y, w, h) to  (x1, y1, x2, y2) ：[https://github.com/Sompote/SAM3_LoRA/blob/5693271e4697e1ec30461549eb90d7148b017516/train_sam3_lora_native.py#L197](url)

But during inference, the model appears to output boxes in (cx, cy, w, h) format：
[https://github.com/Sompote/SAM3_LoRA/blob/5693271e4697e1ec30461549eb90d7148b017516/infer_sam.py#L276](url)

Could this mismatch in coordinate formats between training and inference be causing the large prediction errors?

**Comment by shenxhui** [comment]
> Thanks for reporting this — you're right that there's a box format mismatch, and it is the cause of the bad predictions.
> 
> Root cause:
> 
> SAM3 internally expects target boxes in normalized CxCyWH (center_x, center_y, width, height) format. The back_convert() function, the Hungarian matcher, and all loss functions (L1 + GIoU) all assume this format.
> 
> However, the training dataset code was converting COCO [x, y, w, h] to [x1, y1, x2, y2] (XYXY) and storing that in Object.bbox:
> 
> # Before (wrong) — stored as XYXY
> x, y, w, h = bbox_coco box_tensor = torch.tensor([x, y, x + w, y + h], dtype=torch.float32)
> 
> When back_convert() reads these XYXY values and interprets them as CxCyWH, the targets become garbled. This corrupts:
> 
> * Box regression — L1 loss compares pred CxCyWH against XYXY values misread as CxCyWH
> * GIoU loss — applies box_cxcywh_to_xyxy() on already-XYXY values, producing nonsense coordinates
> * Hungarian matching — wrong matching cascades into wrong mask supervision too
> 
> The inference code was actually correct — it properly interprets model output as CxCyWH. The problem was entirely on the training side.
> 
> Fix:
> 
> Convert COCO boxes to normalized CxCyWH instead of XYXY:
> 
> # After (correct) — stored as CxCyWH
> x, y, w, h = bbox_coco cx = x + w / 2.0 cy = y + h / 2.0 box_tensor = torch.tensor([ cx * scale_w / self.resolution, cy * scale_h / self.resolution, w * scale_w / self.resolution, h * scale_h / self.resolution, ], dtype=torch.float32)
> 
> Also updated the area calculation and validation GT builders to match. This is now fixed. Existing checkpoints trained before this fix will need to be retrained.

Thanks for your clarification!
